rationale/message_block_layout.md

```diff
@@ -0,0 +1,46 @@
+
+We note a nice property: if the transaction sender knows 1) `k`, the size of the matrix, 2) the starting location of their message in a row, and 3) the length of the message (they know this since they are sending the message), then they can actually compute a sequence of roots to _subtrees in the row NMTs_. More importantly, anyone can compute this, and can compute _the simple Merkle root of these subtree roots_.
+
+All transaction senders knowing the exact starting location of their message is unrealistic however, as that would require interaction between the block producer and transaction senders. We can impose some additional rules on message placement to make the possible starting locations of messages sufficiently predictable and constrained such that users can deterministically compute subtree roots without interaction:
+1. Messages begin at a new share, unlike non-message data which can span multiple shares.
+1. Messages that span multiple rows must begin at the start of a row (this can occur if a message is longer than `k` shares _or_ if the block producer decides to start a message partway through a row and it cannot fit).
+1. Messages begin at a location aligned with the largest power of 2 that is not larger than the message length or `k`.
+
+With the above constraints, we can compute subtree roots deterministically easily: simply slice off either the largest power of 2 that isn't larger than the remaining message length, or `k`, whichever is smaller. This is done recursively. As an example, with `k = 4` and message length of `11`, the message would be sliced with lengths `4, 4, 2, 1`. The resulting slices are the leaves of subtrees whose roots can be computed. Due to the rules above, the subtrees are guaranteed to be aligned to powers of 2, and thus the subtree roots will be present as internal nodes in the NMT of _some_ row(s).
+
+This is similar to [Merkle Mountain Ranges](https://www.usenix.org/legacy/event/sec09/tech/full_papers/crosby.pdf), though with the largest subtree bounded by `k` rather than being unbounded.
+
+The last piece of the puzzle is determining _which_ row the message is placed at (or, more specifically, the starting location). This is needed to keep the block producer accountable. To this end, the block producer simply augments each fee-paying transaction with some metadata: the starting location of the message the transaction pays for.
+
+## Caveats
+
+Unfortunately, the message placement rules described above conflict with the first rule that shares must be ordered by namespace ID, as shares between two messages that cannot be placed adjacent to each other do not have a natural namespace they belong to. Assigning these shares a fixed namespace ID disrupts the strict ordering of all shares by namespace ID, but should not meaningfully affect proof sizes into the data NMTs, and so is an acceptable tradeoff.
```

Comment:
Does that not break the security/correctness of NMTs?

Reply:
Fixed in 2e8eb24.